### ISSUE TITLE
AWSDateTime scalar is not currently supported by RDS datetime type

### ISSUE THREAD

**Issue by MaximumCrash** [aws-amplify/amplify-cli#2199]
**Describe the bug**
This is more of an issue with AppSync and RDS, I've already put in a ticket with them and the feature/bug ticket has been put in with the Amazon team. But I thought it'd be good to alert this community on this particular issue. 

When pulling in an RDS datasource and a table has a column of a datetime type, the graphql-transformer applies the AppSync scalar AWSDateTime (as expected). The issue however is that the format of AWSDateTime **(YYYY-MM-DDThh:mm:ss.sssZ)** conflicts with the expected MySQL format of Datetime types **(YYYY-MM-DD HH:MM:SS.sss)**.

My temporary solution is to change all AWSDateTime scalars to strings.

**To Reproduce**
1. Create an RDS cluster, database, and simple table with a column that has a datetime type. 
2. Create a javascript project with amplify. 
3. Add a RDS datasource (MySQL 5.6 Aurora Serverless)
5. See the graphql.schema with AWSDatetime scalars (as expected)
6. Try running a mutation to create a record in the RDS datasource using the expected AWSDateTime format. 
7. See the error returned from RDS

**Expected behavior**
My hope was that RDS datetimes would support the AWSDateTime format out the box, but this is an internal conflict between services.

**Comment by manchicken** [comment]
The JavaScript which is generated also attempts to push dates up in a format different from what `AWSDateTime` expects: `"2019-08-17 05:49:55"`

Omitting the field doesn't help, either, as the resolver seems to choke on the existing value in the table as well.

**Comment by djheru** [comment]
I've also run into this bug, which is present in the example Marketplace schema in the amplify `add-graphql-datasource` section of the docs 😞

**Comment by lspellman** [comment]
I have encountered this issue as well. Existing table data, generated through regular SQL calls, produces a DateTime value in table like this: 2019-04-17-19.00.00 -0500.  When returned via AppSync resolver, the cloudwatch log shows the ResponseMapping transformedTemplate as: 2019-04-18 00:00:00.0 which then results in a null value in the response.

**Comment by mdepascale** [comment]
Our temporary solutions:
**BEFORE**
`$util.qr($context.args.input.put("created_at", $util.time.nowFormatted("yyyy-MM-dd HH:mm:ss")))`

**AFTER**
```
#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.get("created_at") )
     #set( $createdAt = $output.get("created_at").replace(" ", "T"))
     $util.qr($output.put("created_at", "${createdAt}.000Z"))
#end
```
Hope they will solve this issue

**Comment by peterv1988** [comment]
> 
> 
> Our temporary solution:
> **BEFORE**
> `$util.qr($context.args.input.put("created_at", $util.time.nowFormatted("yyyy-MM-dd HH:mm:ss")))`
> 
> **AFTER**
> 
> ```
> #set( $output = $utils.rds.toJsonObject($ctx.result) )
> #if( $output.get("created_at") )
>      #set( $createdAt = $output.get("created_at").replace(" ", "T"))
>      $util.qr($output.put("created_at", "${createdAt}.000Z"))
> #end
> ```
> 
> Hope they will solve this issue

For the less experienced Amplify developers, in which file could I configure this (what seems to be a customer resolver)?

**Comment by mdepascale** [comment]
> For the less experienced Amplify developers, in which file could I configure this (what seems to be a customer resolver)?

Yeah you should copy your `resolver.req.vtl `and `resolver.res.vtl` files from build to your resolver folder, then edit them with whatever you need them to do
They will then become custom resolvers

**Comment by peterv1988** [comment]
> Yeah you should copy your `resolver.req.vtl `and `resolver.res.vtl` files from build to your resolver folder, then edit them with whatever you need them to do
> They will then become custom resolvers

Just to have the solution clear to me and others bumping into this through online search.  I would like to clarify how to implement the solution based on this example schema based, generated from MySQL table:
```
CREATE TABLE IF NOT EXISTS `page` (
  `id` INT NOT NULL AUTO_INCREMENT,
  `createdAt` DATETIME NULL DEFAULT CURRENT_TIMESTAMP,
  `title` VARCHAR(60) NOT NULL,
  PRIMARY KEY (`id`))
ENGINE = InnoDB;
```
It appears that using RDS as datasource with most recent Amplify-CLI the resolvers are already copied to `amplify/backend/api/<API_NAME>/resolvers` after generating schema.

From your example it is not exactly clear to me how to implement the fix, it would be great if you could provide example to assist us on how to the default generated `.res.vtl` files from Amplify should be changed. As it seems to rely on `$utils.toJson(..` to return the output results.

-  `amplify/backend/api/<API_NAME>/resolvers/Query.listPages.res.vtl`, content:
```
$utils.toJson($utils.rds.toJsonObject($ctx.result)[0])
```
-  `amplify/backend/api/<API_NAME>/resolvers/Query.getPage.res.vtl`, content:
```
$utils.toJson($output[0])
```

**Comment by mdepascale** [comment]
Alright assuming your query is 
`getPage(id: Int!) : Page`

you will need to edit 
-  `amplify/backend/api/<API_NAME>/resolvers/Query.putPage.res.vtl` with:
`$util.qr($context.args.input.put("created_at", $util.time.nowFormatted("yyyy-MM-dd HH:mm:ss")))`, just add this line before making the PUT request 
- `amplify/backend/api/<API_NAME>/resolvers/Query.getPage.req.vtl` with:
```
#set( $output = $utils.rds.toJsonObject($ctx.result) )
#if( $output.isEmpty() )
  $util.error("Invalid response from RDS DataSource. See info for the full response.", "InvalidResponse", {}, $output)
#end
#set( $output = $output[0] )
#if( $output.isEmpty() )
  #return
#end
#if( $output.get("created_at") )
     #set( $createdAt = $output.get("created_at").replace(" ", "T"))
     $util.qr($output.put("created_at", "${createdAt}.000Z"))
#end
$utils.toJson($output[0])
```

it should work, I will double check it tomorrow

**Comment by georgeleeventures** [comment]
Is there any update on this issue? I highly doubt the graphql RDS datasource part of Amplify is production-ready. It seems like we'd have to make modifications to all of the resolver templates... Not only do I have this issue with AWSDateTime scalar, but I also have issues with JSON being returned improperly from the resolver too.

**Comment by jaubrey-ebar** [comment]
I confirm this still exists.  It's a pretty ridiculous oversight in the Amplify API IMO.

**Success:** MySql DateTime column in RDS -> GraphQL Result -> Javascript Date
**Failure:** The reverse path

Here's the error I see in the failure scenario, maybe it will help someone realize the issue when searching the web:
`RDSHttp:{"message":"Database error code: 1292. Message: Data truncation: Incorrect datetime value: '2021-10-21T07:21:25.803Z' for column 'modified' at row 1"`
